src/ompl/geometric/planners/experience/src/LRR.cpp

```diff
@@ -0,0 +1,610 @@
+#include "ompl/tools/config/SelfConfig.h"
+#include "ompl/tools/lightning/LightningDB.h"
+
+#include <boost/thread.hpp>
+
+#include <limits>
+
+ompl::geometric::LRR::LRR(const base::SpaceInformationPtr &si, const ompl::tools::LightningDBPtr &experienceDB)
+    : base::Planner(si, "LRR"),
+      experienceDB_(experienceDB),
+      nearestK_(10) // default value
+{
+    specs_.approximateSolutions = true;
+    specs_.directed = true;
+
+    // Repair Planner Specific:
+    repairProblemDef_.reset(new base::ProblemDefinition(si_));
+
+    psk_.reset(new ompl::geometric::PathSimplifier(si_));
+}
+
+ompl::geometric::LRR::~LRR(void)
+{
+    freeMemory();
+}
+
+void ompl::geometric::LRR::clear(void)
+{
+    Planner::clear();
+    freeMemory();
+
+    // Clear the inner planner
+    if (repairPlanner_)
+        repairPlanner_->clear();
+}
+
+void ompl::geometric::LRR::setLightningDB(ompl::tools::LightningDBPtr experienceDB)
+{
+    experienceDB_ = experienceDB;
+}
+
+void ompl::geometric::LRR::setRepairPlanner(const base::PlannerPtr &planner)
+{
+    if (planner && planner->getSpaceInformation().get() != si_.get())
+        throw Exception("Repair planner instance does not match space information");
+    repairPlanner_ = planner;
+    setup_ = false;
+}
+
+void ompl::geometric::LRR::setup(void)
+{
+    Planner::setup();
+
+    // Setup repair planner (for use by the rrPlanner)
+    // Note: does not use the same pdef as the main planner in this class
+    if (!repairPlanner_)
+    {
+        // Set the repair planner
+        boost::shared_ptr<ompl::geometric::RRTConnect> repair_planner( new ompl::geometric::RRTConnect( si_ ) );
+
+        OMPL_DEBUG("No repairing planner specified. Using default: %s", repair_planner->getName().c_str() );
+        repairPlanner_ = repair_planner;
+    }
+
+    // Setup the problem definition for the repair planner
+    repairProblemDef_->setOptimizationObjective(pdef_->getOptimizationObjective()); // copy primary problem def
+
+    // Setup repair planner
+    repairPlanner_->setProblemDefinition(repairProblemDef_);
+    if (!repairPlanner_->isSetup())
+        repairPlanner_->setup();
+}
+
+void ompl::geometric::LRR::freeMemory(void)
+{
+}
+
+ompl::base::PlannerStatus ompl::geometric::LRR::solve(const base::PlannerTerminationCondition &ptc)
+{
+    bool solved = false;
+    bool approximate = false;
+    double approxdif = std::numeric_limits<double>::infinity();
+
+    // Check if the database is empty
+    if (!experienceDB_->getExperiencesCount())
+    {
+        OMPL_INFORM("Experience database is empty so unable to run LRR algorithm.");
+
+        return base::PlannerStatus::CRASH;
+    }
+
+    // Get a single start state TODO: more than one
+    const base::State *startState = pis_.nextStart();
+
+    // Get a single goal state TODO: more than one
+    base::Goal *goal   = pdef_->getGoal().get();
+
+    // Check that we have the correct type of goal
+    if (!goal)
```

Comment:
Actually, that check for ``goal`` is probably unnecessary since the ``goal`` object is never used in the rest of the code. I think it should be removed.

Instead, what should be questioned is the following line:
```
const base::State *goalState = pis_.nextGoal(ptc); 
```

Is that an ok way to get a goal state? LightningRetrieveRepair works by choosing a *single* start and *single* goal state, and searching for a cached path that has the closest start and goal state. It doesn't care about sampleable regions. Is the above line ok? 

I'm commenting out the unnecessary part, at least for now.